src/libraries/System.Security.Cryptography.X509Certificates/tests/PublicKeyTests.cs

```diff
@@ -558,6 +640,149 @@ public static void TestKey_BrainpoolP160r1()
+
+            // Too small
+            Assert.False(key.TryExportSubjectPublicKeyInfo(algSpki.Slice(1), out written), nameof(key.TryExportSubjectPublicKeyInfo));
+            Assert.Equal(0, written);
+        }
+
+        [Fact]
+        public static void CreateFromSubjectPublicKeyInfo_Roundtrip_RSA()
+        {
+            using RSA rsa = RSA.Create();
+            rsa.ImportFromPem(TestData.RsaPkcs8PublicKey);
+            byte[] spki = rsa.ExportSubjectPublicKeyInfo();
+
+            PublicKey key = PublicKey.CreateFromSubjectPublicKeyInfo(spki, out int read);
+
+            Assert.IsAssignableFrom<RSA>(key.Key);
+            Assert.Equal("1.2.840.113549.1.1.1", key.Oid.Value);
+            Assert.Equal(spki, key.ExportSubjectPublicKeyInfo());
+            Assert.Equal(spki.Length, read);
+        }
+
+        [Fact]
+        public static void CreateFromSubjectPublicKeyInfo_Roundtrip_DSA()
+        {
+            using DSA dsa = DSA.Create();
+            dsa.ImportFromPem(TestData.DsaPkcs8PublicKey);
+            byte[] spki = dsa.ExportSubjectPublicKeyInfo();
+
+            PublicKey key = PublicKey.CreateFromSubjectPublicKeyInfo(spki, out int read);
+
+            Assert.IsAssignableFrom<DSA>(key.Key);
+            Assert.Equal("1.2.840.10040.4.1", key.Oid.Value);
+            Assert.Equal(spki, key.ExportSubjectPublicKeyInfo());
+            Assert.Equal(spki.Length, read);
+        }
+
+        [Fact]
+        public static void CreateFromSubjectPublicKeyInfo_Roundtrip_ECDSA()
+        {
+            using ECDsa ecdsa = ECDsa.Create();
+            ecdsa.ImportFromPem(TestData.ECDsaPkcs8PublicKey);
+            byte[] spki = ecdsa.ExportSubjectPublicKeyInfo();
+
+            PublicKey key = PublicKey.CreateFromSubjectPublicKeyInfo(spki, out int read);
+
+            Assert.Throws<NotSupportedException>(() => key.Key);
+            Assert.Equal("1.2.840.10045.2.1", key.Oid.Value);
+            Assert.Equal(spki, key.ExportSubjectPublicKeyInfo());
+            Assert.Equal(spki.Length, read);
+        }
+
+        [Fact]
+        public static void CreateFromSubjectPublicKeyInfo_Roundtrip_ECDH()
+        {
+            using ECDiffieHellman ecdh = ECDiffieHellman.Create();
+            ecdh.ImportFromPem(TestData.EcDhPkcs8PublicKey);
+            byte[] spki = ecdh.ExportSubjectPublicKeyInfo();
+
+            PublicKey key = PublicKey.CreateFromSubjectPublicKeyInfo(spki, out int read);
+
+            Assert.Throws<NotSupportedException>(() => key.Key);
+            Assert.Equal("1.2.840.10045.2.1", key.Oid.Value);
+            Assert.Equal(spki, key.ExportSubjectPublicKeyInfo());
+            Assert.Equal(spki.Length, read);
+        }
+
+        [Fact]
+        public static void CreateFromSubjectPublicKeyInfo_BadEncoding()
+        {
+            Assert.Throws<CryptographicException>(() =>
+                PublicKey.CreateFromSubjectPublicKeyInfo(new byte[] { 0xFF }, out _));
+        }
+
```

Comment:
How about: well formed, but not an algorithm we support, e.g. GOST R 34-10.2001.

```C#
// The SubjectPublicKeyInfo value from the test certificate in
// https://tools.ietf.org/html/rfc4491#section-4.2
const string KeyHex =
    "3063301C06062A8503020213301206072A85030202240006072A850302021E01" +
    "03430004408495687560021A407508CD138C31892CFDE505037A435CF46D2B0F" +
    "E74F327E578FEBCC16B9958803D09A7C85AE0FE48DEAA6BB7E56C7CBB0DF0F66" +
    "BCCAEA1A60";
```

Since it's just doing decode, it seems like it should crack it apart just fine, and put it back together just fine.


Additionally, it'll probably support cracking open and reassembling an invalid DSA key (malformed DSAParameters), but then the Key property will throw.  Probably another good test to have.